src/experimental/dvl.py

```diff
@@ -0,0 +1,298 @@
+            raise Exception()
+
+        self.checks = []
+        if mini is not None:
+            self.checks.append(lambda x: mini <= x)
+        if exclusive_mini is not None:
+            self.checks.append(lambda x: exclusive_mini < x)
+        if maxi is not None:
+            self.checks.append(lambda x: x <= maxi)
+        if exclusive_maxi is not None:
+            self.checks.append(lambda x: x < exclusive_maxi)
+
+    @classmethod
+    def from_string(cls, string):
+        match = cls.single_number.match(string)
+        if match:
+            # need to find out whether it is mini or exclusive_min etc.
+            return cls(mini=int(match.group(1)), maxi=int(match.group(1)))
+
+        match = cls.range_.match(string)
+        mini, exclusive_mini, maxi, exclusive_maxi = None, None, None, None
+        if match:
+            value = match.group(2)
+            if value:
+                if match.group(1) == "(":
+                    exclusive_mini = int(value)
+                elif match.group(1) == "[":
+                    mini = int(value)
+                else:
+                    raise Exception(f"unknown: {match.group(1)}")
+
+            value = match.group(3)
+            if value:
+                if match.group(4) == ")":
+                    exclusive_maxi = int(value)
+                elif match.group(4) == "]":
+                    maxi = int(value)
+                else:
+                    raise Exception(f"unknown: {match.group(4)}")
+
+            return cls(
+                mini=mini,
+                exclusive_mini=exclusive_mini,
+                maxi=maxi,
+                exclusive_maxi=exclusive_maxi,
+            )
+
+        raise Exception(f"unrecognised string {string}")
+
+    def __contains__(self, value):
+        return all(check(value) for check in self.checks)
+
+
+def check_type(type_):
+    def decorator(func):
+        def wrapper(self, instance):
+            if not isinstance(instance, type_):
+                return False
+            return func(self, instance)
+
+        return wrapper
+
+    return decorator
+
+
+class String(Type):
+    def __init__(self, regex=None):
+        self.regex = re.compile(regex) if regex else None
+
+    @check_type(str)
+    def __call__(self, instance):
+        if self.regex:
+            return bool(self.regex.match(instance))
+        return True
+
+
+class Number(Type):
+    def __init__(self, range_=None, multipleOf=None):
+        self.range = Range.from_string(range_) if range_ else None
+        self.multipleOf = multipleOf
+
+    @check_type(numbers.Number)
+    def __call__(self, instance):
+        if self.range and instance not in self.range:
+            return False
+        if self.multipleOf and instance % self.multipleOf != 0:
+            return False
+        return True
+
+
+class Integer(Number):
+    @check_type(int)
+    def __call__(self, instance):
+        return super().__call__(instance)
+
+
+class Bool(Type):
```

Comment:
could just make this a function constructor tbh. that returns the same function. Maybe a micro-optimizaton but who knows.
```python
def check_bool(instance):
    return isinstance(instance, bool)

def Bool():
    return check_bool

def Null():
    return check_null

def any_(instance):
    return True

def Any():
  return any_

def nothing():
    "Matches nothing"
    return False
def Nothing():
    return nothing
```